File: app/src/main/java/com/qalliance/treetracker/TreeTracker/fragments/MapsFragment.java
package com.qalliance.treetracker.TreeTracker.fragments;


import android.Manifest;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.HapticFeedbackConstants;
import android.view.InflateException;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMap.OnMarkerClickListener;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.qalliance.treetracker.TreeTracker.BuildConfig;
import com.qalliance.treetracker.TreeTracker.activities.MainActivity;
import com.qalliance.treetracker.TreeTracker.application.Permissions;
import com.qalliance.treetracker.TreeTracker.R;
import com.qalliance.treetracker.TreeTracker.utilities.ValueHelper;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

public class MapsFragment extends Fragment implements OnClickListener, OnMarkerClickListener, OnMapReadyCallback {
    private static final String TAG = "MapsFragment";

    public interface LocationDialogListener {
		void refreshMap();
	}

	LocationDialogListener mSettingCallback;

	private ArrayList<Marker> redPulsatingMarkers = new ArrayList<Marker>();
	private ArrayList<Marker> redToGreenPulsatingMarkers = new ArrayList<Marker>();

	private SharedPreferences mSharedPreferences;
	private int mCurrentRedToGreenMarkerColor = -1;
	private boolean paused = false;
	protected int mCurrentMarkerColor;
	private static View view;


	public MapsFragment() {
		//some overrides and settings go here
	}

	@Override
	public void onAttach(Context context) {
		super.onAttach(context);

		try {
			mSettingCallback = (LocationDialogListener) context;
		} catch (ClassCastException e) {
			throw new ClassCastException(context.toString()
					+ " must implement LocationDialogListener");
		}
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);


	}

	@Override
	public void onPause() {
		super.onPause();

		paused = true;
		Log.d("GPS_Bugs", "MasFragment onPause");
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		Log.d("GPS_Bugs", "MasFragment on Destroy");
	}

	@Override
	public void onResume() {
		super.onResume();
		Log.d("GPS_Bugs", "MasFragment onResume");
		if (paused) {
			((SupportMapFragment) getChildFragmentManager()
					.findFragmentById(R.id.map)).getMapAsync(this);
		}
		paused = false;

		mCurrentRedToGreenMarkerColor = R.drawable.green_pin;
		mCurrentMarkerColor = R.drawable.red_pin_pulsating_4;

		handler.post(new Runnable() {
			public void run() {
				if (mCurrentRedToGreenMarkerColor == R.drawable.red_pin) {
					mCurrentRedToGreenMarkerColor = R.drawable.green_pin;
				} else {
					mCurrentRedToGreenMarkerColor = R.drawable.red_pin;
				}
				for (Marker marker : redToGreenPulsatingMarkers) {
					marker.setIcon(BitmapDescriptorFactory.fromResource(mCurrentRedToGreenMarkerColor));
				}

				if (!paused)
					handler.postDelayed(this, 500);
			}
		});

		handler.post(new Runnable() {
			public void run() {
				if (mCurrentMarkerColor == R.drawable.red_pin) {
					mCurrentMarkerColor = R.drawable.red_pin_pulsating_1;
				} else if (mCurrentMarkerColor == R.drawable.red_pin_pulsating_1) {
					mCurrentMarkerColor = R.drawable.red_pin_pulsating_2;
				} else if (mCurrentMarkerColor == R.drawable.red_pin_pulsating_2) {
					mCurrentMarkerColor = R.drawable.red_pin_pulsating_3;
				} else if (mCurrentMarkerColor == R.drawable.red_pin_pulsating_3) {
					mCurrentMarkerColor = R.drawable.red_pin_pulsating_4;
				} else if (mCurrentMarkerColor == R.drawable.red_pin_pulsating_4) {
					mCurrentMarkerColor = R.drawable.red_pin;
				}

				for (Marker marker : redPulsatingMarkers) {
					marker.setIcon(BitmapDescriptorFactory.fromResource(mCurrentMarkerColor));
				}

				if (!paused)
					handler.postDelayed(this, 200);
			}
		});
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();

		try {
			SupportMapFragment fragment = (SupportMapFragment) getActivity()
					.getSupportFragmentManager().findFragmentById(
							R.id.map);
			if (fragment != null)
				getActivity().getSupportFragmentManager().beginTransaction().remove(fragment).commit();

		} catch (IllegalStateException e) {
			//handle this situation because you are necessary will get
			//an exception here :-(
		}
	}


	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState) {

		if (view != null) {
			ViewGroup parent = (ViewGroup) view.getParent();
			if (parent != null)
				parent.removeView(view);
		}
		try {
			view = inflater.inflate(R.layout.fragment_map, container, false);
		} catch (InflateException e) {
	        /* map is already there, just return view as it is */
		}

		View v = view;

		mSharedPreferences = getActivity().getSharedPreferences(
				"com.qalliance.treetracker", Context.MODE_PRIVATE);

        if (!((AppCompatActivity) getActivity()).getSupportActionBar().isShowing()) {
            Log.d("MainActivity", "toolbar hide");
            ((AppCompatActivity) getActivity()).getSupportActionBar().show();
        }
		((TextView) getActivity().findViewById(R.id.toolbar_title)).setText(R.string.map);
		((AppCompatActivity)getActivity()).getSupportActionBar().setDisplayHomeAsUpEnabled(false);

        FloatingActionButton fab = (FloatingActionButton) v.findViewById(R.id.fab);
        fab.setOnClickListener(this);

		((SupportMapFragment) getChildFragmentManager()
				.findFragmentById(R.id.map)).getMapAsync(this);


		TextView mapGpsAccuracy = ((TextView) v.findViewById(R.id.fragment_map_gps_accuracy));
		TextView mapGpsAccuracyValue = ((TextView) v.findViewById(R.id.fragment_map_gps_accuracy_value));

		int minAccuracy = mSharedPreferences.getInt(ValueHelper.MIN_ACCURACY_GLOBAL_SETTING, ValueHelper.MIN_ACCURACY_DEFAULT_SETTING);

		if (mapGpsAccuracy != null) {
			Log.i("ođe", "0");
			if (MainActivity.mCurrentLocation != null) {
				Log.i("ođe", "1");
				if (MainActivity.mCurrentLocation.hasAccuracy() && (MainActivity.mCurrentLocation.getAccuracy() < minAccuracy)) {
					Log.i("ođe", "2");
					mapGpsAccuracy.setTextColor(Color.GREEN);
					mapGpsAccuracyValue.setTextColor(Color.GREEN);
					mapGpsAccuracyValue.setText(Integer.toString(Math.round(MainActivity.mCurrentLocation.getAccuracy())) + " " + getResources().getString(R.string.meters));
					MainActivity.mAllowNewTreeOrUpdate = true;
				} else {
					Log.i("ođe", "3");
					mapGpsAccuracy.setTextColor(Color.RED);
					MainActivity.mAllowNewTreeOrUpdate = false;

					if (MainActivity.mCurrentLocation.hasAccuracy()) {
						mapGpsAccuracyValue.setTextColor(Color.RED);
						mapGpsAccuracyValue.setText(Integer.toString(Math.round(MainActivity.mCurrentLocation.getAccuracy())) + " " + getResources().getString(R.string.meters));
					} else {
						mapGpsAccuracyValue.setTextColor(Color.RED);
						mapGpsAccuracyValue.setText("N/A");
					}
				}
			} else {
				Log.i("ođe", "5");
				if (ActivityCompat.checkSelfPermission(getActivity(), android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED &&
						ActivityCompat.checkSelfPermission(getActivity(), android.Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
					requestPermissions(
							new String[]{android.Manifest.permission.ACCESS_COARSE_LOCATION},
							Permissions.MY_PERMISSION_ACCESS_COURSE_LOCATION);
				}
				mapGpsAccuracy.setTextColor(Color.RED);
				mapGpsAccuracyValue.setTextColor(Color.RED);
				mapGpsAccuracyValue.setText("N/A");
				MainActivity.mAllowNewTreeOrUpdate = false;
			}

		}

		return v;
	}

	@Override
	public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
		if (requestCode == Permissions.MY_PERMISSION_ACCESS_COURSE_LOCATION) {
			mSettingCallback.refreshMap();
		}
	}

	private Handler handler = new Handler() {

		@Override
		public void handleMessage(Message msg) {

		}
	};

	private Fragment fragment;

	private Bundle bundle;

	private FragmentTransaction fragmentTransaction;


	public void onClick(View v) {


		v.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY, HapticFeedbackConstants.FLAG_IGNORE_GLOBAL_SETTING);

		Cursor photoCursor;
		switch (v.getId()) {
            case R.id.fab:
            	Log.d(TAG, "fab click");
                if (MainActivity.mAllowNewTreeOrUpdate || BuildConfig.GPS_ACCURACY.equals("off")) {
					fragment = new NewTreeFragment();
					bundle = getActivity().getIntent().getExtras();
					fragment.setArguments(bundle);

					fragmentTransaction = getActivity().getSupportFragmentManager()
							.beginTransaction();
					fragmentTransaction.replace(R.id.container_fragment, fragment).addToBackStack(ValueHelper.NEW_TREE_FRAGMENT).commit();
				} else {
					Toast.makeText(getActivity(), "Insufficient GPS accuracy.", Toast.LENGTH_SHORT).show();
				}
				break;
//			case R.id.fragment_map_update_tree:
//
//				if (MainActivity.mAllowNewTreeOrUpdate) {
//					SQLiteDatabase db = MainActivity.dbHelper.getReadableDatabase();
//
////					String query = "select * from tree_photo " +
////							"left outer join tree on tree._id = tree_id " +
////							"left outer join photo on photo._id = photo_id " +
////							"left outer join location on location._id = photo.location_id " +
////							"where is_outdated = 'N'";
//
//					String query = "select * from tree " +
//							"left outer join location on location._id = tree.location_id " +
//							"left outer join tree_photo on tree._id = tree_id " +
//							"left outer join photo on photo._id = photo_id ";
//
//					Log.e("query", query);
//
//					photoCursor = db.rawQuery(query, null);
//
//					if (photoCursor.getCount() <= 0) {
//						Toast.makeText(getActivity(), "No trees to update", Toast.LENGTH_SHORT).show();
//						db.close();
//						return;
//					}
//
//					db.close();
//
//					fragment = new UpdateTreeFragment();
//					bundle = getActivity().getIntent().getExtras();
//					fragment.setArguments(bundle);
//
//					fragmentTransaction = getActivity().getSupportFragmentManager()
//							.beginTransaction();
//					fragmentTransaction.replace(R.id.container_fragment, fragment).addToBackStack(ValueHelper.UPDATE_TREE_FRAGMENT).commit();
//				} else {
//					Toast.makeText(getActivity(), "Insufficient GPS accuracy.", Toast.LENGTH_SHORT).show();
//				}
//
//				break;
		}


	}

	public boolean onMarkerClick(Marker marker) {
		fragment = new TreePreviewFragment();
		bundle = getActivity().getIntent().getExtras();

		if (bundle == null)
			bundle = new Bundle();

		bundle.putString(ValueHelper.TREE_ID, marker.getTitle());
		fragment.setArguments(bundle);

		fragmentTransaction = getActivity().getSupportFragmentManager()
				.beginTransaction();
		fragmentTransaction.replace(R.id.container_fragment, fragment).addToBackStack(ValueHelper.TREE_PREVIEW_FRAGMENT).commit();
		return true;
	}


	@Override
	public void onMapReady(GoogleMap map) {


		if (ActivityCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
			// TODO: Consider calling
			//    ActivityCompat#requestPermissions
			// here to request the missing permissions, and then overriding
			//   public void onRequestPermissionsResult(int requestCode, String[] permissions,
			//                                          int[] grantResults)
			// to handle the case where the user grants the permission. See the documentation
			// for ActivityCompat#requestPermissions for more details.
			return;
		}
		map.setMyLocationEnabled(true);

		SQLiteDatabase db = MainActivity.dbHelper.getReadableDatabase();

		Cursor treeCursor = db.rawQuery("select *, tree._id as tree_id from tree left outer join location on location_id = location._id where is_missing = 'N'", null);
		treeCursor.moveToFirst();

		redToGreenPulsatingMarkers.clear();
		redPulsatingMarkers.clear();

		if (treeCursor.getCount() > 0) {

			LatLng latLng = new LatLng(-33.867, 151.206);
			int bla = 0;

			do {


				Log.e("time_for_update", treeCursor.getString(treeCursor.getColumnIndex("_id")));

				SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");


				Boolean isSynced = Boolean.parseBoolean(treeCursor.getString(treeCursor.getColumnIndex("is_synced")));

				Log.e("issynced", Boolean.toString(isSynced));

				Date dateForUpdate = new Date();
				try {
					dateForUpdate = dateFormat.parse(treeCursor.getString(treeCursor.getColumnIndex("time_for_update")));
				} catch (ParseException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}

				Date updated = new Date();
				try {
					updated = dateFormat.parse(treeCursor.getString(treeCursor.getColumnIndex("time_updated")));
				} catch (ParseException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}

				Date created = new Date();
				try {
					created = dateFormat.parse(treeCursor.getString(treeCursor.getColumnIndex("time_created")));
				} catch (ParseException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}

				boolean priority = treeCursor.getString(treeCursor.getColumnIndex("is_priority")).equals("Y");
				latLng = new LatLng(Double.parseDouble(treeCursor.getString(treeCursor.getColumnIndex("lat"))),
						Double.parseDouble(treeCursor.getString(treeCursor.getColumnIndex("long"))));


				MarkerOptions markerOptions = new MarkerOptions()
						.title(Long.toString(treeCursor.getLong(treeCursor.getColumnIndex("tree_id"))))// set Id instead of title
						.icon(BitmapDescriptorFactory.fromResource(R.drawable.green_pin))
						.position(latLng);
				Marker marker = map.addMarker(markerOptions);



				if (priority) {
					redPulsatingMarkers.add(marker);
					continue;
				}


				Log.i("updated", "*************");
				Log.i("updated", updated.toLocaleString());
				Log.i("dateForUpdate", dateForUpdate.toLocaleString());


				if (dateForUpdate.before(new Date())) {


					Log.e("updated", "should be red");

					Calendar calendar = Calendar.getInstance();
					calendar.setTime(dateForUpdate);

					Calendar currCalendar = Calendar.getInstance();
					currCalendar.setTime(new Date());

					marker.setIcon(BitmapDescriptorFactory.fromResource(R.drawable.red_pin));
				}

//		        Log.i("updated", "*************");
//		        if (created.before(updated) && !isSynced) {
//		        	redToGreenPulsatingMarkers.add(marker);
//		        }
			} while (treeCursor.moveToNext());


			db.close();

			map.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 20));

		} else {
			if (MainActivity.mCurrentLocation != null) {
				LatLng myLatLng = new LatLng(MainActivity.mCurrentLocation.getLatitude(), MainActivity.mCurrentLocation.getLongitude());
				map.moveCamera(CameraUpdateFactory.newLatLngZoom(myLatLng, 10));
			}
		}

		map.setOnMarkerClickListener(MapsFragment.this);

		// Other supported types include: MAP_TYPE_NORMAL,
		// MAP_TYPE_TERRAIN, MAP_TYPE_HYBRID and MAP_TYPE_NONE
		map.setMapType(GoogleMap.MAP_TYPE_NORMAL);
	}
}

File: app/src/main/java/org/greenstand/android/TreeTracker/fragments/MapsFragment.kt
package org.greenstand.android.TreeTracker.fragments

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.Group
import androidx.core.app.ActivityCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.maps.android.clustering.ClusterManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import org.greenstand.android.TreeTracker.R
import org.greenstand.android.TreeTracker.database.TreeTrackerDAO
import org.greenstand.android.TreeTracker.databinding.FragmentMapBinding
import org.greenstand.android.TreeTracker.map.TreeMapMarker
import org.greenstand.android.TreeTracker.models.FeatureFlags
import org.greenstand.android.TreeTracker.models.LocationUpdateManager
import org.greenstand.android.TreeTracker.models.Planter
import org.greenstand.android.TreeTracker.utilities.ImageUtils
import org.greenstand.android.TreeTracker.utilities.TreeClusterRenderer
import org.greenstand.android.TreeTracker.utilities.mainActivity
import org.greenstand.android.TreeTracker.utilities.vibrate
import org.greenstand.android.TreeTracker.viewmodels.MapViewModel
import org.koin.android.ext.android.inject
import org.koin.android.viewmodel.ext.android.viewModel
import timber.log.Timber

class MapsFragment :
    androidx.fragment.app.Fragment(),
    OnClickListener,
    OnMapReadyCallback,
    View.OnLongClickListener {

    private lateinit var bindings: FragmentMapBinding

    private val vm: MapViewModel by viewModel()
    private val locationUpdateManager: LocationUpdateManager by inject()
    private val dao: TreeTrackerDAO by inject()
    private val user: Planter by inject()

    private var mapFragment: SupportMapFragment? = null
    private var map: GoogleMap? = null

    private lateinit var convergenceProgressView: Group
    private lateinit var clusterManager: ClusterManager<TreeMapMarker>

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        bindings = FragmentMapBinding.inflate(inflater)
        return bindings.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        convergenceProgressView = view.findViewById<Group>(R.id.convergenceProgressGroup)
        convergenceProgressView.visibility = View.GONE

        vm.checkInStatusLiveData.observe(
            this,
            Observer<Boolean> { planterIsCheckedIn ->
                if (planterIsCheckedIn) {
                    GlobalScope.launch(Dispatchers.Main) {

                       mainActivity().bindings.toolbarTitle.text = vm.getPlanterName()

                        val photoPath = user.profilePhotoPath
                        val profileImageView = bindings.mapUserImage

                        if (photoPath != null) {
                            val rotatedBitmap =
                                ImageUtils.decodeBitmap(photoPath, resources.displayMetrics.density)
                            if (rotatedBitmap != null) {
                                profileImageView.setImageBitmap(rotatedBitmap)
                                profileImageView.visibility = View.VISIBLE
                            }
                        } else {
                            profileImageView.visibility = View.GONE
                        }
                    }
                }
            }
        )

        if (mapFragment == null) {
            mapFragment = SupportMapFragment()
            childFragmentManager.beginTransaction().apply {
                add(R.id.map, mapFragment!!)
                commit()
            }
        }

        if (!(activity as AppCompatActivity).supportActionBar!!.isShowing) {
            Timber.d("toolbar hide")
            (activity as AppCompatActivity).supportActionBar!!.show()
        }

        (activity as AppCompatActivity).supportActionBar!!.setDisplayHomeAsUpEnabled(false)

        bindings.addTreeButton.setOnClickListener(this)

        bindings.goToUploadsButton.setOnClickListener(this)

        if (FeatureFlags.DEBUG_ENABLED) {
            bindings.addTreeButton.setOnLongClickListener(this)
        }

        mapFragment!!.getMapAsync(this)
    }

    override fun onClick(v: View) {
        v.vibrate()
        when (v.id) {
            R.id.addTreeButton -> {
                Timber.d("fab click")
                // Disable the addTreeButton below to avoid triggering the onClick listener
                // more than one once.
                bindings.addTreeButton.isEnabled = false
                lifecycleScope.launch {
                    if (vm.requiresLogin()) {
                        findNavController().navigate(
                            MapsFragmentDirections.actionGlobalLoginFlowGraph()
                        )
                    } else {
                        vm.turnOnTreeCaptureMode()
                        convergenceProgressView.visibility = View.VISIBLE
                        vm.resolveLocationConvergence()
                        convergenceProgressView.visibility = View.GONE
                        findNavController()
                            .navigate(MapsFragmentDirections.actionMapsFragmentToNewTreeGraph())
                    }
                }
            }
            R.id.goToUploadsButton -> {
//                findNavController().navigate(
//                    MapsFragmentDirections.actionMapsFragmentToDataFragment(
//                        true
//                    )
//                )
            }
        }
    }

    // For debug analysis purposes only
    @SuppressLint("SimpleDateFormat")
    override fun onLongClick(view: View): Boolean {
        GlobalScope.launch(Dispatchers.Main) {
            Toast.makeText(activity, "Adding 500 trees", Toast.LENGTH_LONG).show()
            val didSuceed = withContext(Dispatchers.IO) { vm.createFakeTrees() }
            if (didSuceed) {
                Toast.makeText(activity, "500 trees added", Toast.LENGTH_LONG).show()
                renderTrees()
            } else {
                Toast.makeText(activity, "Error adding test trees", Toast.LENGTH_LONG).show()
            }
        }
        return true
    }

    @SuppressLint("MissingPermission")
    override fun onPause() {
        super.onPause()
        map?.isMyLocationEnabled = false
    }

    override fun onResume() {
        super.onResume()

        GlobalScope.launch {
            vm.checkForValidUser()
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        map = googleMap.apply {
            uiSettings.isMyLocationButtonEnabled = false
            uiSettings.isMapToolbarEnabled = false
        }

        clusterManager = ClusterManager(context, map)
        clusterManager.renderer = TreeClusterRenderer(requireContext(), googleMap, clusterManager)
        clusterManager.setOnClusterItemClickListener {
            findNavController().navigate(
                MapsFragmentDirections.actionMapsFragmentToTreePreviewFragment(
                    it.title
                )
            )
            true
        }
        map!!.setOnCameraIdleListener(clusterManager)
        map!!.setOnMarkerClickListener(clusterManager)

        if (ActivityCompat.checkSelfPermission(context!!, Manifest.permission.ACCESS_FINE_LOCATION)
            != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                    context!!,
                    Manifest.permission.ACCESS_COARSE_LOCATION
                ) != PackageManager.PERMISSION_GRANTED
        ) {
            // TODO: Consider calling
            //    ActivityCompat#requestPermissions
            // here to request the missing permissions, and then overriding
            //   public void onRequestPermissionsResult(int requestCode, String[] permissions,
            //                                          int[] grantResults)
            // to handle the case where the user grants the permission. See the documentation
            // for ActivityCompat#requestPermissions for more details.
            return
        }
        map!!.isMyLocationEnabled = true

        map!!.mapType = GoogleMap.MAP_TYPE_NORMAL
        renderTrees()
    }

    private fun renderTrees() {

        map ?: return

        runBlocking {
            val trees = withContext(Dispatchers.IO) { dao.getTreeDataForMap() }

            clusterManager.clearItems()

            if (trees.isNotEmpty()) {
                Timber.d("Adding markers")
                for (tree in trees) {
                    val treeMapAnnotation = TreeMapMarker(
                        tree.latitude,
                        tree.longitude,
                        _title = tree.treeCaptureId.toString()
                    )
                    clusterManager.addItem(treeMapAnnotation)
                }
            }
            if (locationUpdateManager.currentLocation != null) {
                val myLatLng = LatLng(
                    locationUpdateManager.currentLocation!!.latitude,
                    locationUpdateManager.currentLocation!!.longitude
                )
                map!!.moveCamera(CameraUpdateFactory.newLatLngZoom(myLatLng, 10f))
            }
        }
    }
}
